grudge/trace_pair.py

```diff
@@ -275,58 +280,49 @@ def connected_ranks(dcoll: DiscretizationCollection):
 class _RankBoundaryCommunication:
     base_tag = 1273
 
-    def __init__(self, dcoll: DiscretizationCollection,
-                 remote_rank, vol_field, tag=None):
-        self.tag = self.base_tag
-        if tag is not None:
-            self.tag += tag
-
-        self.dcoll = dcoll
-        self.array_context = vol_field.array_context
-        self.remote_btag = BTAG_PARTITION(remote_rank)
-        self.bdry_discr = dcoll.discr_from_dd(self.remote_btag)
+    def __init__(self,
+                 dcoll: DiscretizationCollection,
+                 array_container: ArrayOrContainerT,
+                 remote_rank, tag=None):
+        actx = get_container_context_recursively(array_container)
+        btag = BTAG_PARTITION(remote_rank)
 
-        from grudge.op import project
+        local_array_ctr = project(dcoll, "vol", btag, array_container)
+        numpy_data = to_numpy(flatten(local_array_ctr, actx), actx)
+        comm = dcoll.mpi_communicator
 
-        self.local_dof_array = project(dcoll, "vol", self.remote_btag, vol_field)
+        self.dcoll = dcoll
+        self.array_context = actx
+        self.remote_btag = btag
+        self.bdry_discr = dcoll.discr_from_dd(btag)
+        self.local_array_ctr = local_array_ctr
 
-        local_data = self.array_context.to_numpy(flatten(self.local_dof_array))
-        comm = self.dcoll.mpi_communicator
+        self.tag = self.base_tag
+        if tag is not None:
+            self.tag += tag
 
-        self.send_req = comm.Isend(local_data, remote_rank, tag=self.tag)
-        self.remote_data_host = np.empty_like(local_data)
-        self.recv_req = comm.Irecv(self.remote_data_host, remote_rank, self.tag)
+        self.send_req = comm.Isend(numpy_data, remote_rank, tag=self.tag)
```

Comment:
This might be a good moment to look up mpi4py semantics (and add a comment on our usage here). As it's written, `numpy_data` will go out of scope shortly after the `Isend` is posted, and it's not obvious whether that's long enough to allow the send to complete.

My quick skim of the [mpi4py docs](https://mpi4py.readthedocs.io/en/stable/overview.html#nonblocking-communications) doesn't settle this unambiguously, unfortunately.

At any rate, I think the safe approach might be to keep the buffer alive until the send completes (while adding a comment summarizing what we know/assume about the lifetime semantics).

Reply:
I've gone ahead provided some comments/notes (to the best of my understanding): https://github.com/inducer/grudge/pull/154/commits/55950b121a97fcca77e73c356dfa62bba501b3ca